<a href="https://colab.research.google.com/github/kylemcq13/Projects/blob/master/16_8_Feature_Engineering_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment

Say, we want to understand the relationship between the expenditures of the governments and the students' overall success in the math and reading.

1. Create a new score variable from the weighted averages of all score variables in the datasets. **Notice that the number of students in the 4th grade isn't the same as the number of students in the 8th grade. So, you should appropriately weigh the scores!**.
2. What are the correlations between this newly created score variable and the expenditure types? Which 1 of the expenditure types is more correlated than the others?
3. Now, apply PCA to the 4 expenditure types. How much of the total variance is explained by the 1st component?
4. What is the correlation between the overall score variable and the 1st principal component? 
5. If you were to choose the best variables for your model, would you prefer using the 1st principal component instead of the expenditure variables? Why?

In [0]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sqlalchemy import create_engine
import warnings
import math
from matplotlib.mlab import PCA as mlabPCA
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

warnings.filterwarnings('ignore')
sns.set(style="whitegrid")

In [0]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'useducation'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

edu_df = pd.read_sql_query('select * from useducation',con=engine)

engine.dispose()

In [0]:
#interpolate the data as before

columns = ["ENROLL", "TOTAL_REVENUE", "FEDERAL_REVENUE", 
           "STATE_REVENUE", "LOCAL_REVENUE", "TOTAL_EXPENDITURE", 
           "INSTRUCTION_EXPENDITURE", "SUPPORT_SERVICES_EXPENDITURE",
           "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE", "GRADES_PK_G", 
           "GRADES_KG_G", "GRADES_4_G", "GRADES_8_G", "GRADES_12_G", "GRADES_1_8_G", 
           "GRADES_9_12_G", "GRADES_ALL_G", "AVG_MATH_4_SCORE", "AVG_MATH_8_SCORE", 
           "AVG_READING_4_SCORE", "AVG_READING_8_SCORE"]

edu5_df = edu_df.copy()
states = edu5_df["STATE"].unique()

for col in columns:
    for state in states:
        edu5_df.loc[edu5_df["STATE"] == state, col] = edu5_df.loc[edu5_df["STATE"] == state, col].fillna(edu5_df[edu5_df["STATE"] == state][col].interpolate())
        
edu5_df.dropna(inplace=True)
edu5_df

PRIMARY_KEY                 STATE  YEAR     ENROLL  \
54                 1993_ARIZONA               ARIZONA  1993   676297.0   
55                1993_ARKANSAS              ARKANSAS  1993   311432.0   
57                1993_COLORADO              COLORADO  1993   539538.0   
58             1993_CONNECTICUT           CONNECTICUT  1993   471918.0   
59                1993_DELAWARE              DELAWARE  1993   104355.0   
60    1993_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1993    80937.0   
61                 1993_FLORIDA               FLORIDA  1993  1980569.0   
106                1994_ARIZONA               ARIZONA  1994   708482.0   
107               1994_ARKANSAS              ARKANSAS  1994   312695.0   
109               1994_COLORADO              COLORADO  1994   549570.0   
110            1994_CONNECTICUT           CONNECTICUT  1994   479522.0   
111               1994_DELAWARE              DELAWARE  1994   103782.0   
113   1994_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1994    80678.0   
114                1994_FLORIDA               FLORIDA  1994  2039884.0   
159                1995_ARIZONA               ARIZONA  1995   735018.0   
160               1995_ARKANSAS              ARKANSAS  1995   447757.0   
162               1995_COLORADO              COLORADO  1995   639805.0   
163            1995_CONNECTICUT           CONNECTICUT  1995   489513.0   
164               1995_DELAWARE              DELAWARE  1995   106813.0   
165   1995_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1995    80450.0   
166                1995_FLORIDA               FLORIDA  1995  2107514.0   
209                1996_ALABAMA               ALABAMA  1996   735912.0   
211                1996_ARIZONA               ARIZONA  1996   764681.0   
212               1996_ARKANSAS              ARKANSAS  1996   452907.0   
213             1996_CALIFORNIA            CALIFORNIA  1996  5460484.0   
214               1996_COLORADO              COLORADO  1996   655679.0   
215            1996_CONNECTICUT           CONNECTICUT  1996   499815.0   
216               1996_DELAWARE              DELAWARE  1996   108461.0   
217   1996_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1996    79802.0   
218                1996_FLORIDA               FLORIDA  1996  2175352.0   
...                         ...                   ...   ...        ...   
1462                  2017_IOWA                  IOWA  2017   508014.0   
1463                2017_KANSAS                KANSAS  2017   495545.0   
1464              2017_KENTUCKY              KENTUCKY  2017   686440.0   
1465             2017_LOUISIANA             LOUISIANA  2017   660561.0   
1466                 2017_MAINE                 MAINE  2017   179879.0   
1467              2017_MARYLAND              MARYLAND  2017   879196.0   
1468         2017_MASSACHUSETTS         MASSACHUSETTS  2017   921029.0   
1469              2017_MICHIGAN              MICHIGAN  2017  1335713.0   
1470             2017_MINNESOTA             MINNESOTA  2017   811157.0   
1471           2017_MISSISSIPPI           MISSISSIPPI  2017   486245.0   
1472              2017_MISSOURI              MISSOURI  2017   891554.0   
1473               2017_MONTANA               MONTANA  2017   145240.0   
1474              2017_NEBRASKA              NEBRASKA  2017   315542.0   
1475                2017_NEVADA                NEVADA  2017   441623.0   
1476         2017_NEW_HAMPSHIRE         NEW_HAMPSHIRE  2017   179682.0   
1477            2017_NEW_JERSEY            NEW_JERSEY  2017  1364473.0   
1478            2017_NEW_MEXICO            NEW_MEXICO  2017   319861.0   
1479              2017_NEW_YORK              NEW_YORK  2017  2590945.0   
1480        2017_NORTH_CAROLINA        NORTH_CAROLINA  2017  1462036.0   
1481          2017_NORTH_DAKOTA          NORTH_DAKOTA  2017   108384.0   
1482                  2017_OHIO                  OHIO  2017  1595024.0   
1483             2017_TENNESSEE             TENNESSEE  2017   994530.0   
1484                 201

# 1. Create a new score variable from the weighted averages of all score variables in the datasets. Notice that the number of students in the 4th grade isn't the same as the number of students in the 8th grade. So, you should appropriately weigh the scores!.

In [0]:
#using a weight factor of 0.5 for each grade, weight the scores based on student population of each grade

edu5_df['weighted_scores'] = (edu5_df['GRADES_4_G'] * ((edu5_df['AVG_MATH_4_SCORE'] + edu5_df['AVG_READING_4_SCORE']) * 0.5) + edu5_df['GRADES_8_G'] * ((edu5_df['AVG_MATH_8_SCORE'] + edu5_df['AVG_READING_8_SCORE']) * 0.5)) / (edu5_df['GRADES_4_G'] + edu5_df['GRADES_8_G'])
edu5_df

PRIMARY_KEY                 STATE  YEAR     ENROLL  \
54                 1993_ARIZONA               ARIZONA  1993   676297.0   
55                1993_ARKANSAS              ARKANSAS  1993   311432.0   
57                1993_COLORADO              COLORADO  1993   539538.0   
58             1993_CONNECTICUT           CONNECTICUT  1993   471918.0   
59                1993_DELAWARE              DELAWARE  1993   104355.0   
60    1993_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1993    80937.0   
61                 1993_FLORIDA               FLORIDA  1993  1980569.0   
106                1994_ARIZONA               ARIZONA  1994   708482.0   
107               1994_ARKANSAS              ARKANSAS  1994   312695.0   
109               1994_COLORADO              COLORADO  1994   549570.0   
110            1994_CONNECTICUT           CONNECTICUT  1994   479522.0   
111               1994_DELAWARE              DELAWARE  1994   103782.0   
113   1994_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1994    80678.0   
114                1994_FLORIDA               FLORIDA  1994  2039884.0   
159                1995_ARIZONA               ARIZONA  1995   735018.0   
160               1995_ARKANSAS              ARKANSAS  1995   447757.0   
162               1995_COLORADO              COLORADO  1995   639805.0   
163            1995_CONNECTICUT           CONNECTICUT  1995   489513.0   
164               1995_DELAWARE              DELAWARE  1995   106813.0   
165   1995_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1995    80450.0   
166                1995_FLORIDA               FLORIDA  1995  2107514.0   
209                1996_ALABAMA               ALABAMA  1996   735912.0   
211                1996_ARIZONA               ARIZONA  1996   764681.0   
212               1996_ARKANSAS              ARKANSAS  1996   452907.0   
213             1996_CALIFORNIA            CALIFORNIA  1996  5460484.0   
214               1996_COLORADO              COLORADO  1996   655679.0   
215            1996_CONNECTICUT           CONNECTICUT  1996   499815.0   
216               1996_DELAWARE              DELAWARE  1996   108461.0   
217   1996_DISTRICT_OF_COLUMBIA  DISTRICT_OF_COLUMBIA  1996    79802.0   
218                1996_FLORIDA               FLORIDA  1996  2175352.0   
...                         ...                   ...   ...        ...   
1462                  2017_IOWA                  IOWA  2017   508014.0   
1463                2017_KANSAS                KANSAS  2017   495545.0   
1464              2017_KENTUCKY              KENTUCKY  2017   686440.0   
1465             2017_LOUISIANA             LOUISIANA  2017   660561.0   
1466                 2017_MAINE                 MAINE  2017   179879.0   
1467              2017_MARYLAND              MARYLAND  2017   879196.0   
1468         2017_MASSACHUSETTS         MASSACHUSETTS  2017   921029.0   
1469              2017_MICHIGAN              MICHIGAN  2017  1335713.0   
1470             2017_MINNESOTA             MINNESOTA  2017   811157.0   
1471           2017_MISSISSIPPI           MISSISSIPPI  2017   486245.0   
1472              2017_MISSOURI              MISSOURI  2017   891554.0   
1473               2017_MONTANA               MONTANA  2017   145240.0   
1474              2017_NEBRASKA              NEBRASKA  2017   315542.0   
1475                2017_NEVADA                NEVADA  2017   441623.0   
1476         2017_NEW_HAMPSHIRE         NEW_HAMPSHIRE  2017   179682.0   
1477            2017_NEW_JERSEY            NEW_JERSEY  2017  1364473.0   
1478            2017_NEW_MEXICO            NEW_MEXICO  2017   319861.0   
1479              2017_NEW_YORK              NEW_YORK  2017  2590945.0   
1480        2017_NORTH_CAROLINA        NORTH_CAROLINA  2017  1462036.0   
1481          2017_NORTH_DAKOTA          NORTH_DAKOTA  2017   108384.0   
1482                  2017_OHIO                  OHIO  2017  1595024.0   
1483             2017_TENNESSEE             TENNESSEE  2017   994530.0   
1484                 201

# 2. What are the correlations between this newly created score variable and the expenditure types? Which 1 of the expenditure types is more correlated than the others?

In [0]:
#created a correlation table of weighted scores and expenditures

edu5_df[['weighted_scores','TOTAL_EXPENDITURE', 'INSTRUCTION_EXPENDITURE', 'SUPPORT_SERVICES_EXPENDITURE', 'OTHER_EXPENDITURE']].corr()

weighted_scores  TOTAL_EXPENDITURE  \
weighted_scores                      1.000000           0.093020   
TOTAL_EXPENDITURE                    0.093020           1.000000   
INSTRUCTION_EXPENDITURE              0.105787           0.990829   
SUPPORT_SERVICES_EXPENDITURE         0.098643           0.993175   
OTHER_EXPENDITURE                    0.027635           0.946483   

                              INSTRUCTION_EXPENDITURE  \
weighted_scores                              0.105787   
TOTAL_EXPENDITURE                            0.990829   
INSTRUCTION_EXPENDITURE                      1.000000   
SUPPORT_SERVICES_EXPENDITURE                 0.975867   
OTHER_EXPENDITURE                            0.907235   

                              SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  
weighted_scores                                   0.098643           0.027635  
TOTAL_EXPENDITURE                                 0.993175           0.946483  
INSTRUCTION_EXPENDITURE                           0.975867           0.907235  
SUPPORT_SERVICES_EXPENDITURE                      1.000000           0.953789  
OTHER_EXPENDITURE                                 0.953789           1.000000

It seems as if the instruction expenditure has the most correlation with the weighted scores.

# 3. Now, apply PCA to the 4 expenditure types. How much of the total variance is explained by the 1st component?

In [0]:
#calculate the PCA of the 4 expenditure types

pca_exp = edu5_df[['INSTRUCTION_EXPENDITURE', 'TOTAL_EXPENDITURE', 'SUPPORT_SERVICES_EXPENDITURE', 'OTHER_EXPENDITURE']]

pca_exp = StandardScaler().fit_transform(pca_exp)

sklearn_pca = PCA(n_components=1)
edu5_df['pca_col'] = sklearn_pca.fit_transform(pca_exp)

print(sklearn_pca.explained_variance_ratio_)

[0.97107935]


According to the calculation, over 97% of the variance is explained by the 1st component.

# 4. What is the correlation between the overall score variable and the 1st principal component?

In [0]:
#produce a correlation table

edu5_df[['weighted_scores', 'pca_col', 'INSTRUCTION_EXPENDITURE', 'TOTAL_EXPENDITURE', 'SUPPORT_SERVICES_EXPENDITURE', 'OTHER_EXPENDITURE']].corr()

weighted_scores   pca_col  \
weighted_scores                      1.000000  0.082805   
pca_col                              0.082805  1.000000   
INSTRUCTION_EXPENDITURE              0.105787  0.983098   
TOTAL_EXPENDITURE                    0.093020  0.997324   
SUPPORT_SERVICES_EXPENDITURE         0.098643  0.995350   
OTHER_EXPENDITURE                    0.027635  0.965638   

                              INSTRUCTION_EXPENDITURE  TOTAL_EXPENDITURE  \
weighted_scores                              0.105787           0.093020   
pca_col                                      0.983098           0.997324   
INSTRUCTION_EXPENDITURE                      1.000000           0.990829   
TOTAL_EXPENDITURE                            0.990829           1.000000   
SUPPORT_SERVICES_EXPENDITURE                 0.975867           0.993175   
OTHER_EXPENDITURE                            0.907235           0.946483   

                              SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  
weighted_scores                                   0.098643           0.027635  
pca_col                                           0.995350           0.965638  
INSTRUCTION_EXPENDITURE                           0.975867           0.907235  
TOTAL_EXPENDITURE                                 0.993175           0.946483  
SUPPORT_SERVICES_EXPENDITURE                      1.000000           0.953789  
OTHER_EXPENDITURE                                 0.953789           1.000000

As can be seen from the table above, the correlation between the first component and weighted score variable is 0.0828 or roughtly 8.3%.

# 5. If you were to choose the best variables for your model, would you prefer using the 1st principal component instead of the expenditure variables? Why?

According to our table, the expenditures are more correlated with our weighted scores than the principal component. Of the expenditures, instruction expenditure is the most correlated so it would make sense to use this. Also, according to the literature, PCA works best on variables that are correlated to 0.8 or lower. In our case, the expenditures correlations all exceed this value which could lead to problems. 